src/engines/outlookcalendar.ts

```diff
@@ -0,0 +1,125 @@
+import * as msal from "@azure/msal-node";
+import * as graph from "@microsoft/microsoft-graph-client";
+import * as dateFns from "date-fns";
+import "isomorphic-fetch";
+
+import { getUnixTime } from "../util";
+
+const SCOPES = ["calendars.readwrite", "mailboxsettings.read", "user.read"];
+
+let auth: graph.Client | undefined;
+
+interface EventDate {
+  dateTime: string;
+  timeZone: string;
+}
+
+interface Event {
+  body: {
+    content: string;
+    contentType: string;
+  };
+  bodyPreview: string;
+  end: EventDate;
+  id: string;
+  lastModifiedDateTime: string;
+  organizer: {
+    emailAddress: {
+      address: string;
+      name: string;
+    };
+  };
+  start: EventDate;
+  subject: string;
+  webLink: string;
+}
+
+const engine: Engine = {
+  id: "outlookcalendar",
+  init: ({
+    clientId,
+    clientSecret,
+    token,
+  }: {
+    clientId: string;
+    clientSecret: string;
+    token: string;
+  }) => {
+    // Instantiate msal application object
+    const msalClient = new msal.ConfidentialClientApplication({
+      auth: {
+        authority: "https://login.microsoftonline.com/common/",
+        clientId,
+        clientSecret,
+      },
+    });
+
+    // Initialize Graph client
+    auth = graph.Client.init({
+      // Implement an auth provider that gets a token from the app's MSAL
+      // instance
+      authProvider: async (
+        done: (error: any, accessToken: string | null) => void,
+      ) => {
+        try {
+          // Use refresh token to acquire new access token
+          const auth_result = await msalClient.acquireTokenByRefreshToken({
+            refreshToken: token,
+            scopes: SCOPES,
+          });
+
+          if (auth_result) {
+            // First param to callback is the error,
+            // Set to null in success case
+            done(null, auth_result.accessToken);
+          }
+        } catch (err) {
+          console.log(JSON.stringify(err, Object.getOwnPropertyNames(err)));
+          done(err, null);
+        }
+      },
+    });
+  },
+  name: "Outlook Calendar",
+  search: async q => {
+    if (!auth) {
+      throw Error("Engine not initialized");
+    }
+
+    const today = new Date();
+    const events = await auth
+      .api("/me/calendarview")
+      // Add the begin and end of the calendar window (20 year window, past and future)
+      .query({
+        endDateTime: dateFns.addDays(today, 912).toDateString(),
+        startDateTime: dateFns.addDays(today, -912).toDateString(),
```

Comment:
Looks like this is actually a 5 year window. I've never used Outlook Calendar so I'm not sure what range it supports or what range you intend to use

Reply:
I originally intended to use an arbitrary 20-year window, but the API allows only a maximum of a 5-year window. I forgot to update the comment. I will update it.